### PR TITLE
Closes #1607, Closes #1678: Change disable_login_form functionality, disable login form by default, add link to webauth to login form.

### LINKED ISSUE #1607: Login form is missing a link to use CAS
## Problem/Motivation
On default UAQS installs, the login form was set to automatically redirect to CAS login.

On default AZQS installs, the login form shows the normal Drupal login with no CAS option, which can be confusing for users trying to log into the site with their NetID.

Per #1170, the merged solution was to provide an option in AZ Quickstart settings to disable the standard /user/login path (and /user login redirect by association). But this setting is not enabled by default, leaving the standard login form available for standard installs. Since CAS is enabled by default, an option to get to the CAS login should be added to the login form.

Also related, users may be used to going to /user or /user/login as was available in UAQS sites. The solution to #1170 doesn't sufficiently address that user expectation when enabled.

Ideally, instead of disabling the form entirely and making it an Access Denied page, the setting should instead either hide the form and leave the WebAuth link on the page (as recommended by the CAS module) or redirect to CAS login as was the experience in UAQS.

### Describe the bug
The Drupal login page is available by default on new AZQS installs. The CAS login option is missing from the login form.

## Proposed resolution
The CAS WebAuth login option should be made available on the login form.
Enable "Place a link to log in via CAS on the standard /user/login form" in CAS settings by default.

### Expected behavior
A link to login with WebAuth should show up on the login form.

Alternatively, if CAS is intended to be the primary method of logins for deployed AZQS sites, follow the CAS module's guidance:
> If CAS is the primary way your users will log in, it is recommended to alter the login page in a custom module to hide the standard form.

Instead of disabling the page entirely as with the current AZQS AZ CAS setting option, change to hiding/removing the form or redirect to CAS login.

### Current login form
image

### Suggested login form
image

### LINKED ISSUE #1678: Disable default user login form by default
## Motivation
Follow-up to #1170.  In #1173 we added a new setting to the az_cas module for disabling the default Drupal login form but we did not enable the setting by default because it broke many of our existing automated tests.

## Proposed Resolution
We should update our tests to work with the disable login form setting enabled by default.

### PR DESCRIPTION
**Pull request by joeparsons**
## Description

Alternate PR for addressing both #1607 and #1678 together.  Replaces and combines #1667 and #1679:

- Moved `disable_login_form` implementation from route subscriber to `az_cas_form_user_login_form_alter()` (per suggestions from @trackleft and @joshuasosa). 
- Simplified existing code in `az_cas_form_user_login_form_alter()` (from #1667) to only set the weight of the login link.
- Turn on `disable_login_form` setting by default (from #1679).
- Updates existing tests to prevent failures caused by new default disable_login_form setting (from #1679)
- These needed to be combined because both use `az_cas_form_user_login_form_alter()`.
- Updates existing AZ CAS settings test to work with new `disable_login_form` implementation.

## Related issues

- Closes #1607 
- Closes #1678

### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_cas/az_cas.module [hunk 1]
```diff
@@ -69,3 +70,49 @@ function az_cas_validate_roles(array &$form, FormStateInterface $form_state) {
...
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Modify the user login form to put CAS login below Drupal login.
+ */
+function az_cas_form_user_login_form_alter(&$form, FormStateInterface $form_state) {
+  $config = \Drupal::config('cas.settings');
+  $az_config = \Drupal::config('az_cas.settings');
```
Comment (joeparsons) [inline modules/custom/az_cas/az_cas.module:81]:
```suggestion
  $cas_config = \Drupal::config('cas.settings');
  $az_cas_config = \Drupal::config('az_cas.settings');
```

File: modules/custom/az_cas/az_cas.module [hunk 2]
```diff
@@ -69,3 +70,49 @@ function az_cas_validate_roles(array &$form, FormStateInterface $form_state) {
...
+
+  // Cached form must be busted if we alter AZ CAS settings.
+  $form['#cache']['tags'] = array_merge($form['#cache']['tags'], $az_config->getCacheTags());
```
Comment (joeparsons) [inline modules/custom/az_cas/az_cas.module:84]:
```suggestion
  $form['#cache']['tags'] = array_merge($form['#cache']['tags'], $az_cas_config->getCacheTags());
```

File: modules/custom/az_cas/az_cas.module [hunk 3]
```diff
@@ -69,3 +70,49 @@ function az_cas_validate_roles(array &$form, FormStateInterface $form_state) {
...
+
+  // Add a link to login via CAS on the user login form.
+  $login_link_enabled = $config->get('login_link_enabled');
```
Comment (joeparsons) [inline modules/custom/az_cas/az_cas.module:87]:
```suggestion
  $login_link_enabled = $cas_config->get('login_link_enabled');
```

File: modules/custom/az_cas/az_cas.module [hunk 4]
```diff
@@ -69,3 +70,49 @@ function az_cas_validate_roles(array &$form, FormStateInterface $form_state) {
...
+  if ($login_link_enabled) {
+    $form['cas_login_link']['#weight'] = 100;
+  }
+
+  $disable_login_form = $az_config->get('disable_login_form');
```
Comment (joeparsons) [inline modules/custom/az_cas/az_cas.module:92]:
```suggestion
  $disable_login_form = $az_cas_config->get('disable_login_form');
```

File: modules/custom/az_cas/az_cas.module [hunk 5]
```diff
@@ -69,3 +70,49 @@ function az_cas_validate_roles(array &$form, FormStateInterface $form_state) {
...
+ * Modify the user login form to put CAS login below Drupal login.
+ */
+function az_cas_form_user_login_form_alter(&$form, FormStateInterface $form_state) {
+  $cas_config = \Drupal::config('cas.settings');
+  $az_cas_config = \Drupal::config('az_cas.settings');
+
+  // Cached form must be busted if we alter AZ CAS settings.
+  $form['#cache']['tags'] = array_merge($form['#cache']['tags'], $az_cas_config->getCacheTags());
+
+  // Add a link to login via CAS on the user login form.
```
Comment (joeparsons) [inline modules/custom/az_cas/az_cas.module:86]: // Adjust the weight of the CAS login link (if enabled).

Comment (joeparsons) [inline modules/custom/az_cas/az_cas.module:86]:
```suggestion
  // Adjust the weight of the CAS login link (if enabled). 
```

**Comment by joeparsons** [comment]
If this PR is merged, @joshuasosa should be included as a co-author.

**Review by trackleft — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Pull request merged**